### PR TITLE
Add ability to configure the product name and version telemetry

### PR DESCRIPTION
**Pull request by clguimanMSFT**
## Description
Add ability to configure the product information sent as part of telemetry.
If this feature is not used, the data is identical with before.
If the feature is used, then the product name & version will be amended with '/' + custom information. I don't have any preferences for the separator, if you have a better idea, I'm ok with changing it.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x ] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [ x] Title of the pull request is clear and informative.
- [x ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [x ] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [x ] If applicable, the public code is properly documented.
- [ ] Pull request includes test coverage for the included changes. (I can't verify that the telemetry is being sent. I assume it was working before and my change shouldn't brake it)
- [x] The code builds without any errors.

### CODE REVIEW

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ClientInfo.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package com.microsoft.azure.eventhubs;
+
+import com.microsoft.azure.eventhubs.amqp.AmqpConstants;
+import org.apache.qpid.proton.amqp.Symbol;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ClientInfo {
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ClientInfo.java]: add Javadoc - a guidance statement like - generic frameworks like yours or other famous examples - are the candidates who should use this.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ClientInfo.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    private String productName = ClientConstants.PRODUCT_NAME;
+
+    private String productVersion = ClientConstants.CURRENT_JAVACLIENT_VERSION;
+
+    public static ClientInfo Default = new ClientInfo(null, null);
+
+    public ClientInfo(final String productName, final String productVersion) {
+        if (productName != null) {
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ClientInfo.java]: nit: consider this pattern if you like :)

private final String productName;
public ClientInfo(final String productName, final String productVersion) { 
this.productName = productName == null ? constants.Product_name : ...;

and then getConnProperties just initialize once lazily - whenever it gets called first.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventHubClient.java [hunk 3]
```diff
@@ -53,7 +53,21 @@ private EventHubClient(final ConnectionStringBuilder connectionString) throws IO
...
+    /**
+     * Synchronous version of {@link #createFromConnectionString(String)}.
+     *
+     * @param connectionString The connection string to be used. See {@link ConnectionStringBuilder} to construct a connectionString.
+     * @param clientInfo       A custom {@link ClientInfo} to be sent to EventHub
+     * @return EventHubClient which can be used to create Senders and Receivers to EventHub
+     * @throws EventHubException If Service Bus service encountered problems during connection creation.
+     * @throws IOException         If the underlying Proton-J layer encounter network errors.
+     */
+    public static EventHubClient createFromConnectionStringSync(final String connectionString, final ClientInfo clientInfo)
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventHubClient.java]: adding the ClientInfo to this façade method of our client might make us look complicated - with too many options for the starters (and then, we will also need to replicate this across all of our clients):

public static EventHubClient createFromConnectionStringSync(final String connectionString, final ClientInfo clientInfo) 

let me discuss with our pack of client developers and come back on the best way to pass this.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ClientInfo.java (hunk 1 above)
Comment (clguimanMSFT) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ClientInfo.java]: Is this fine?
/**
 * {@link ClientInfo} can be used to append the user's product information onto the metrics collected by the SDK
 *
 * Usage example:
 * AzureEventHub Logstash Input plugin (https://github.com/Azure/azure-diagnostics-tools/tree/master/Logstash/logstash-input-azureeventhub)
 */

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ClientInfo.java (hunk 2 above)
Comment (clguimanMSFT) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/ClientInfo.java]: I didn't get the 'getConnProperties' suggestion. Should I cache the map as a class member and initialize it the first time the method is called?

**Comment by SreeramGarlapati** [comment]
@clguimanMSFT - I couldn't get hold of our developers today - will update you regarding this on 9/14.

**Comment by clguimanMSFT** [comment]
After talking offline I simplified the feature implementation and created a new PR: https://github.com/Azure/azure-event-hubs-java/pull/173
Closing this one

**Pull request closed without merging**